File: examples/go/celsius/go/src/celsius/celsius.go
// Copyright 2017 The Wallaroo Authors.
//
//  Licensed under the Apache License, Version 2.0 (the "License");
//  you may not use this file except in compliance with the License.
//  You may obtain a copy of the License at
//
//      http://www.apache.org/licenses/LICENSE-2.0
//
//  Unless required by applicable law or agreed to in writing, software
//  distributed under the License is distributed on an "AS IS" BASIS,
//  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or
//  implied. See the License for the specific language governing
//  permissions and limitations under the License.

package main

import (
  "bytes"
  "C"
  "encoding/binary"
  "encoding/gob"
  "flag"
  "fmt"
  "math"
  "reflect"
    "strings"
  wa "wallarooapi"
  app "wallarooapi/application"
)

//export ApplicationSetup
func ApplicationSetup(showHelp bool) *C.char {
  fs := flag.NewFlagSet("wallaroo", flag.ExitOnError)
  inHostsPortsArg := fs.String("in", "", "input host:port list")
  outHostsPortsArg := fs.String("out", "", "output host:port list")

  fs.Parse(wa.Args[1:])

  inHostsPorts := hostsPortsToList(*inHostsPortsArg)

  inHost := inHostsPorts[0][0]
  inPort := inHostsPorts[0][1]

  outHostsPorts := hostsPortsToList(*outHostsPortsArg)
  outHost := outHostsPorts[0][0]
  outPort := outHostsPorts[0][1]

  wa.Serialize = Serialize
  wa.Deserialize = Deserialize

  application := app.MakeApplication("Celsius to Fahrenheit")
  application.NewPipeline("Celsius Conversion", app.MakeTCPSourceConfig(inHost, inPort, &Decoder{})).
    To(&MultiplyBuilder{}).
    To(&AddBuilder{}).
    ToSink(app.MakeTCPSinkConfig(outHost, outPort, &Encoder{}))

  return C.CString(application.ToJson())
}

type MultiplyBuilder struct {}

func (rb *MultiplyBuilder) Build() interface{} {
  return &Multiply{}
}

type Multiply struct {}

func (r *Multiply) Name() string {
  return "multiply by 1.8"
}

func (r *Multiply) Compute(data interface{}) interface{} {
  input := (data.(float32))

  return input * 1.8
}

type AddBuilder struct {}

func (rb *AddBuilder) Build() interface{} {
  return &Add{}
}

type Add struct {}

func (r *Add) Name() string {
  return "add 32"
}

func (r *Add) Compute(data interface{}) interface{} {
  input := (data.(float32))

  return input + 32.0
}

type Decoder struct {}

func (d *Decoder) HeaderLength() uint64 {
  return 4
}

func (d *Decoder) PayloadLength(b []byte) uint64 {
  return uint64(binary.BigEndian.Uint32(b[0:4]))
}

func (d* Decoder) Decode(b []byte) interface{} {
  return float32FromByteSlice(b)
}

type Encoder struct {}

func (e *Encoder) Encode(data interface{}) []byte {
  msg := data.(float32)
  s := fmt.Sprintf("%.6f", msg)
  return []byte(s + "\n")
}

const (
  float32Type = iota
  multiplyType
  multiplyBuilderType
  addType
  addBuilderType
  decoderType
  encoderType
)

func Serialize(c interface{}) []byte {
  switch t := c.(type) {
  case float32:
    buff := make([]byte, 4)
    binary.BigEndian.PutUint32(buff, float32Type)
    var b bytes.Buffer
    enc := gob.NewEncoder(&b)
    enc.Encode(c)
    return append(buff, b.Bytes()...)
  case *Multiply:
     buff := make([]byte, 4)
    binary.BigEndian.PutUint32(buff, multiplyType)
    return buff
  case *MultiplyBuilder:
    buff := make([]byte, 4)
    binary.BigEndian.PutUint32(buff, multiplyBuilderType)
    return buff
  case *Add:
     buff := make([]byte, 4)
    binary.BigEndian.PutUint32(buff, addType)
    return buff
  case *AddBuilder:
    buff := make([]byte, 4)
    binary.BigEndian.PutUint32(buff, addBuilderType)
    return buff
  case *Decoder:
    buff := make([]byte, 4)
    binary.BigEndian.PutUint32(buff, decoderType)
    return buff
  case *Encoder:
    buff := make([]byte, 4)
    binary.BigEndian.PutUint32(buff, encoderType)
    return buff
  default:
    fmt.Println("SERIALIZE MISSED A CASE")
    fmt.Println(reflect.TypeOf(t))
  }

  return nil
}

func Deserialize(buff []byte) interface{} {
  componentType := binary.BigEndian.Uint32(buff[:4])
  payload := buff[4:]

  switch componentType {
  case float32Type:
    b := bytes.NewBuffer(payload)
    dec := gob.NewDecoder(b)
    var f float32
    dec.Decode(&f)
    return f
  case multiplyType:
    return &Multiply{}
  case multiplyBuilderType:
    return &MultiplyBuilder{}
  case addType:
    return &Add{}
  case addBuilderType:
    return &AddBuilder{}
  case decoderType:
    return &Decoder{}
  case encoderType:
    return &Encoder{}
  default:
    fmt.Println("DESERIALIZE MISSED A CASE")
  }

  return nil
}

func hostsPortsToList(hostsPorts string) [][]string {
  hostsPortsList := make([][]string, 0)
  for _, hp := range strings.Split(hostsPorts, ",") {
    hostsPortsList = append(hostsPortsList, strings.Split(hp, ":"))
  }
  return hostsPortsList
}

func float32FromByteSlice(buff []byte) float32 {
  bits := binary.BigEndian.Uint32(buff)
  float := math.Float32frombits(bits)

  return float
}

func main() {
}

File: examples/python/celsius_connectors/celsius.py
# Copyright 2017 The Wallaroo Authors.
#
#  Licensed under the Apache License, Version 2.0 (the "License");
#  you may not use this file except in compliance with the License.
#  You may obtain a copy of the License at
#
#      http://www.apache.org/licenses/LICENSE-2.0
#
#  Unless required by applicable law or agreed to in writing, software
#  distributed under the License is distributed on an "AS IS" BASIS,
#  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or
#  implied. See the License for the specific language governing
#  permissions and limitations under the License.

"""
This is an example of a stateless application that takes a floating point
Celsius value and sends out a floating point Fahrenheit value.
"""

import struct

import wallaroo
import wallaroo.experimental


def application_setup(args):
    out_host, out_port = wallaroo.tcp_parse_output_addrs(args)[0]
    in_host, in_port = wallaroo.tcp_parse_input_addrs(args)[0]
    #source_config = wallaroo.TCPSourceConfig("celsius", in_host, in_port, decoder)
    source_config = wallaroo.experimental.SourceConnectorConfig(
        "celsius_feed",
        encoder=encode_feed,
        decoder=decode_feed,
        host=in_host,
        port=in_port,
        cookie="Dragons Love Tacos!",
        max_credits=10,
        refill_credits=8)
    #sink_config = wallaroo.experimental.SinkConnectorConfig(
    #    "fahrenheit_conversion",
    #    encoder=encode_conversion,
    #    decoder=decode_conversion,
    #    port=7200)
    sink_config = wallaroo.TCPSinkConfig(out_host, out_port, encode_conversion)
    pipeline = (
        wallaroo.source("convert temperature readings", source_config)
        .to(multiply)
        .to(add)
        .to_sink(sink_config)
    )
    return wallaroo.build_application("Celsius to Fahrenheit", pipeline)


@wallaroo.computation(name="multiply by 1.8")
def multiply(data):
    return data * 1.8


@wallaroo.computation(name="add 32")
def add(data):
    return data + 32


@wallaroo.experimental.stream_message_encoder
def encode_feed(data):
    return data


@wallaroo.experimental.stream_message_decoder
def decode_feed(data):
    return struct.unpack(">f", data)[0]

@wallaroo.decoder(4, ">I")
def decoder(data):
    print("nh: {!r}".format(data))
    return struct.unpack(">f", data)[0]

#@wallaroo.encoder
@wallaroo.experimental.stream_message_encoder
def encode_conversion(data):
    return "{}\n".format(data).encode('utf-8')


@wallaroo.experimental.stream_message_decoder
def decode_conversion(data):
    return data.decode('utf-8')
